transmogrifier/sources/transformer.py

```diff
@@ -362,70 +328,98 @@ def record_is_deleted(cls, source_record: dict[str, JSON] | Tag) -> bool:
             source_record: A single source record.
         """
 
-    def get_optional_fields(self, _source_record: dict[str, JSON] | Tag) -> dict | None:
+    @final
+    def get_optional_field_methods(self) -> Iterator[tuple[str, Callable]]:
         """
-        Retrieve optional TIMDEX fields from a source record.
+        Return optional TIMDEX field names and corresponding methods.
 
-        May be overridden by source subclasses.
+        May not be overridden.
+        """
+        for field_name in [
+            field_name
+            for field_name in timdex.TimdexRecord.get_optional_field_names()
+            if hasattr(self, f"get_{field_name}")
+        ]:
+            yield field_name, getattr(self, f"get_{field_name}")
 
-        Args:
-            source_record: A single source record.
+    @final
+    def generate_derived_fields(
+        self, timdex_record: timdex.TimdexRecord
+    ) -> timdex.TimdexRecord:
```

Comment:
This method gives me a little pause, though I feel like it's conceptually what we want: some kind of operation that writes field values, shared by all transformers, based on values set by the transformers.

What feels kind of unique, unexpected, is that it **modifies** the `timdex_record` instance directly.  By comparison:
1. for required fields, we see them explicitly set during `timdex_record` init
2. for optional fields, we use `setattr()` to set fields based on the outputs of field methods

What if... this method _yielded_ values that we want to set in the timdex record?  If so, we could set them in a "visible" fashion much like we do for optional fields.  And theoretically, we could also add similar `try/except` logic for each.

For the optional fields loop we know the field name and get a value, so can set it:
```python
for field_name, field_method in self.get_optional_field_methods():
    setattr(timdex_record, field_name, field_method(source_record))
```

What if we did something similar for "dynamic" fields?
```python
for field_name, field_value in self.generate_derived_fields(timdex_record):
    setattr(timdex_record, field_name, field_value)
```

This would mean this method generate_derived_fields() would yield field name + field value each time it determines it should.

Zooming all the way back to `transform()`, I feel like it sets up a record building pattern:
1. init record with required fields
2. build up with optional
3. finalize with dynamic

Reply:
Not sure if this would be easier to test, but it feels like it might?  Any given part of `generate_derived_fields` we could test with a `list(...)` and see the values we _would_ set on the `TimdexRecord`.

Reply:
Totally correct, this breaks the pattern and the optional fields pattern is very appropriate, the `source_record` would just become the `timdex_record` in these methods. And yes it will be easier to test, the tests I had for this were more involved than necessary

Reply:
The only complication vs. `optional_fields` is that we need to check if `dates` and `locations` were already set and which would determine whether we `setattr` for new attributes or `extend` the existing attributes. That shouldn't be too difficult though because I was already doing it in the first draft of these methods

Reply:
Deduping is another issue, this is starting to look more complicated than I initially hoped